### ISSUE TITLE
[BUG] Using page.route with postData size > 7.4k fails on Firefox

### ISSUE THREAD

**Issue by jhyot** [microsoft/playwright#16027]
**Context:**
- Playwright Version: [what Playwright version do you use?] 1.22.0 (Java)
- Operating System: [e.g. Windows, Linux or Mac] Mac
- Node.js version: [e.g. 12.22, 14.6] Using Java runner
- Browser: [e.g. All, Chromium, Firefox, WebKit] Firefox
- Extra: [any specific details about your environment]

**Code Snippet**
Java:
```java
  page.route(VERSIONS_API_PATH, route -> {
      if (!route.request().method().equals("PATCH")) {
        route.resume();
        return;
      }

      var data = "a".repeat(7500); // this to simulate large payload

      route.resume(
          new Route.ResumeOptions()
              .setPostData(data));
    });
```

**Describe the bug**

After being processed by the route shown above, Firefox will only really send the first 7436 characters of the postdata to the server. Chromium and Webkit work correctly and send the full payload.

NB, I'm not sure if this is for playwright core or playwright-java.

If you tell me what debug options to enable, I can run the test and provide the debug logs.

**Comment by yury-s** [comment]
@jhyot I couldn't reproduce it neither in node js nor in java. Could it be that the server is trimming the request? If not, we'll need a full reproducible example.

Js snippet:
```js
  test('should send large post data', async ({ page, server }) => {
    await page.goto(server.EMPTY_PAGE);
    const data = 'a'.repeat(7500);
    await page.route('**/*', route => {
      route.continue({ postData: data });
    });
    const [serverRequest] = await Promise.all([
      server.waitForRequest('/sleep.zzz'),
      page.evaluate(() => fetch('/sleep.zzz', { method: 'PATCH', body: 'birdy' }))
    ]);
    expect((await serverRequest.postBody).toString('utf8')).toBe(data);
  });
```

Java snippet:
```java
  @Test
  void shouldSendLargePostData() throws ExecutionException, InterruptedException {
    page.navigate(server.EMPTY_PAGE);
    StringBuilder builder = new StringBuilder();
    for (int i = 0; i < 1 << 14; i++) {
      builder.append("a");
    }
    String data = builder.toString();
    page.route("**/*", route -> {
      route.resume(new Route.ResumeOptions().setPostData(data));
    });
    Future<Server.Request> serverRequest = server.futureRequest("/sleep.zzz");
    page.evaluate("() => fetch('/sleep.zzz', { method: 'PATCH', body: 'birdy' })");
    assertEquals(data, new String(serverRequest.get().postBody, UTF_8));
  }
```

**Comment by jhyot** [comment]
I cannot reproduce it with the simplest bare-bone example.
I think it shouldn't be the server's fault, because Chromium and Webkit are working correctly. I will keep trying to find out what is the cause, and will report my findings here.

**Comment by jhyot** [comment]
I think I found out the root cause.

There is one problem with Firefox, which I left out of my example above:
Firefox only (not Chromium or Webkit) will set the Content-Type header to `application/octet-stream` whenever the postData is set in the route. If I don't use `new Route.ResumeOptions().setPostData(data)`, the original Content-Type ist used (i.e. `application/json` in my case).

Because of the wrong Content-Type, I also modified the headers in the route:
```
      var headers = route.request().allHeaders();
      headers.put("content-type", "application/json");

      route.resume(
          new Route.ResumeOptions()
              .setHeaders(headers)
              .setPostData(data));
```

The problem here is that this method left the Content-Length as the original value, which of course does not match with the modified body, and the server cut off the HTTP request after the original length.

Let me know if I should open a new bug for the Firefox Content-Type behaviour, or if this is somehow expected.
